RubyConf-NG/opensource
# #173833675 Add user resource with roles

#### What does this PR do?
Adds a user resource with roles
#### Description of Task to be completed?
* Users can add doctors
* Users can add patients
* Add tests for creating patients
* Add tests for creating doctors
* Set up roles with rolify
* Setup `guard` for rerunning tests on file change
* Setup RSpec
#### How should this be manually tested?
* Run `rails s` to start your server
* To create a patient, send:

```
curl -L -X POST 'http://localhost:3000/v1/patients' \
-H 'Content-Type: application/json' \
--data-raw '{
    "name": "Kachi",
    "email": "kachi.patient@kachi.com",
    "phone_number": "0102",
    "gender": "male",
    "date_of_birth": "1993-01-01",
    "password": "password",
    "password_confirmation": "password"
}'
```

Which should return:

```.json
{
    "data": {
        "patient": {
            "id": "user-8b4e094881714f79266f2a4e78d5b7dc",
            "name": "Kachi",
            "email": "kachi.patient@kachi.com",
            "phone_number": "0102",
            "date_of_birth": "1993-01-01",
            "gender": "male",
            "roles": [
                "Patient"
            ]
        }
    },
    "extra": null
}
```

* To create a doctor, send:

```
curl -L -X POST 'http://localhost:3000/v1/doctors' \
-H 'Content-Type: application/json' \
--data-raw '{
    "name": "Kachi",
    "email": "kachi.doctor@kachi.com",
    "phone_number": "0102",
    "gender": "male",
    "date_of_birth": "1993-01-01",
    "password": "password",
    "password_confirmation": "password"
}'
```

Which should return:

```.json
{
    "data": {
        "doctor": {
            "id": "user-8b4e094881714f79266f2a4e78d5b7dc",
            "name": "Kachi",
            "email": "kachi.doctor@kachi.com",
            "phone_number": "0102",
            "date_of_birth": "1993-01-01",
            "gender": "male",
            "roles": [
                "Doctor"
            ]
        }
    },
    "extra": null
}
```

#### Any background context you want to provide?
* I used `rolify` - https://github.com/RolifyCommunity/rolify
* I used `activerecord-postgres_enum` - https://github.com/bibendi/activerecord-postgres_enum (Postgres has issues with enums and migrations containing them would cause schema.rb to crash without showing an error. Tests would fail and takes a while to trace it. This gem helps with that)
* More info on the above - https://blog.testdouble.com/posts/2019-06-03-enumerate-your-enums/
* I added a `.rubocop.yml` to help with the conventions we decide to go with. It's pretty skinny and the major rule in it is me preferring double quotes
* I added an error handler I use in my own projects. This can change
* I used `active_model_serializer` - https://github.com/rails-api/active_model_serializers . We could decide to roll our own serializer down the line
* I used a simple versioning system for the controllers. Except `application_controller`, all other controllers are in the `v1` folder under controllers
* I make use of `uids` as the primary IDs' that are shown to the user. So instead of `users/1`, you'd have `users/user-hggf1hdh34`
* I added a simple response method in `app/controllers/concerns/v1/response.rb`. This could be changed too
* I set up `guard`. I use it to watch for file changes and restarts tests. To use, run `bundle exec guard` or `guard`

#### What are the relevant pivotal tracker stories? (if applicable)
https://www.pivotaltracker.com/story/show/173833675
